athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/handlers/MetadataHandler.java

```diff
@@ -446,6 +458,37 @@ public void onPing(PingRequest request)
         //NoOp
     }
 
+    /**
+     * Helper function to get the AmazonS3 instance of the lambda
+     */
+    public AmazonS3 getS3()
+    {
+        return AmazonS3ClientBuilder.standard().build();
+    }
+
+    /**
+     * Helper function used to check if the account that calls the lambda function owns the spill bucket
+     */
+    @VisibleForTesting
+    void checkBucketAuthZ()
+    {
+        if (bucketState == BucketState.VALID) {
+            return;
+        }
+        else if (bucketState == BucketState.INVALID) {
+            throw new RuntimeException("You do NOT own the spill bucket with the name: " + spillBucket);
+        }
+
+        AmazonS3 s3 = getS3();
+        Set<String> buckets = s3.listBuckets().stream().map(b -> b.getName()).collect(Collectors.toSet());
+        if (!buckets.contains(spillBucket)) {
+            bucketState = BucketState.INVALID;
+            throw new RuntimeException("You do NOT own the spill bucket with the name: " + spillBucket);
+        }
+
+        bucketState = BucketState.VALID;
```

Comment:
Not all customers will want to use spill, we should only enforce this if they actually set a spill bucket. I also don't see where we are re-checking if the bucket changed. Instead of storing just the state, I think you should store both the state and the bucket that you checked. If the bucket changes you need to re-check it.

Additionally, I think we should extract this logic into a separate class called SpillLocationVerifier.

Reply:
pretty sure env var change causes instance re-creation so should be no need to re-check

Reply:
That is 100% true for Lambda but this code wont always run in Lambda :)

Reply:
@avirtuos Can you give me a scenario as to how the spill bucket would be changed? 

Moving to a new class is a good idea, this would make unit tests cleaner and more accessible

Reply:
customers can update the spill bucket any time they want. I'm not sure if this answers your question though so please elaborate if you need more context.

Reply:
As far as I know, when an instance of any MetadataHandler child class is initialized, there's no way to update the spillbucket variable in the base class. So I'm not sure how a customer can change the spillbucket. Or, do you mean a customer can somehow change the SPILL_BUCKET_ENV after initialization? So in that case, any change in SPILL_BUCKET_ENV should be checked before performing getSplits?

Reply:
Ah, i see what you mean. Fair point. My request still remains, the cache should be per-spill location. If the spill location changes, the cache of the validity check should expire. The core reason behind this request is that a future maintainer of the configuration (e.g. how we set the spill bucket) has no way to know that other parts of the code (e.g. your check) depend on the value of spill bucket never changing. Its a simple change to have the cache track the spill bucket value so my suggestion is for us to be defensive about it.

Reply:
OK, thanks for the clarification, I'll push the new changes soon.